Modify volume mounts SELinux labels on the fly based on :Z or :z

rhatdan opened the pull request:
Looking for comment on this patch.  

The first patch should be able to be applied without any controversy.  The send patch is extending the qualifiers on the -v command to allow an admin to tell the system to relabel, content.  There might be a need for something similar for changing the DAC Permissions.

vbatts commented:
when testing locally, something in the docker-py tests is failing. Though looking at drone details, it's unclear.

vbatts commented:
ha, now it's green. /me takes another look.

rhatdan commented:
Yes I fixed the patch to still provide VolumesRW, which docker-py has hard coded in its tests.
I was hoping to replace this with VolumesMode to show the mode string specified on mount, but now we have both.

cpuguy83 commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -1785,15 +1785,28 @@ network and environment of the `redis` container via environment variables.
+Docker to relabel file objects on the shared volumes. The 'z' option tells
+Docker that the volume content will be shared between containers. Docker will
+label the content with a shared content label. Shared volumes labels allow all
+containers to read/write content. The 'Z' option tells Docker to label the
```
I know I've gone back and forth on this, but again...

This basically means if I want to bind-mount a dir, I can have selinux make sure that no other containers can read/write from that dir?

rhatdan replied:
"Z" means no other non-priviliged containers could read/write the bind mount.  "z" means one of more containers can read/write the bind mount (From an SELinux point of view.  No 'z' or 'Z' means it depends on the label of the mount point whether or not a container can read and or write the mount point.  Most likely not.

rhatdan commented:
One of the comments earlier was about rewriting Mode to be a structure rather then a string.  I would be willing to do this, if the API/CLI was approved.

cpuguy83 commented:
So anytime there are 2 containers that need to share a volume, it must have a "z" (either both specifying the same bindmount or volumes-from).

Seems like there is a bug (or maybe just too much control?) here where someone can "--volumes-from" and overwrite an existing label. What probably needs to happen is someone should always specify `--volumes-from someone:z`, but they could do `--volumes-from someone:Z` and make the volume unwritable to the other container.

rhatdan commented:
No this is just a problem for volume mounts --volume-from should already be handled.  That had a separate fix.  --volumes-from should not accept the Z or the z option.

cpuguy83 commented:
@rhatdan It does accept z and Z (tested on a CentOS 7 host)

rhatdan commented:
Most likely this is a bug, and I believe the option does nothing. Probably just says it is valid since it is using the same processing as -v /foo:/bar.

cpuguy83 commented on daemon/volumes.go:
```diff
@@ -333,7 +334,7 @@ func (container *Container) setupMounts() error {
 			Source:      container.Volumes[path],
 			Destination: path,
-			Writable:    container.VolumesRW[path],
+			Mode:        container.VolumesMode[path],
```
Instead of passing these mode values directly (and subsequently pulling in github.com/docker/volumes into libcontainer to parse, maybe we can split this out, use "Writable", "LabelType" (shared/unshared/nil)

cpuguy83 commented on vendor/src/github.com/docker/libcontainer/mount/mount.go:
```diff
@@ -7,6 +7,7 @@ import (
 	"syscall"
 
 	"github.com/docker/docker/pkg/symlink"
+	"github.com/docker/docker/volumes"
```
We probably shouldn't import this into libcontainer.

rhatdan commented:
How about we just create a mode object and then put the code in it to handle the verification.
Then the mode object could be in libcontainer, and used by docker.

rhatdan commented:
Rewrote this patch to use a struct Mode, which eliminates the volumes being used in libcontainer.
It is not working, and I am not sure why.  @crosbymichael The Mode gets setup correctly in docker but by the time it gets to Mount in libcontainer it dropped its data.  Is there something special I have to do about passing struct data between docker and libcontainer.  Almost looks like a new instances gets created.

rhatdan commented:
@crosbymichael I guess I am still waiting for patches to get updated for libcontainer?

crosbymichael commented:
@rhatdan we merged the libcontainer changes into docker, we can look at your mode PR again and get this one moving if you want.

rhatdan commented:
Ok, but we need to go back to the old one which passed around Mode/relabel as a string.

rhatdan commented:
This patch is still broken, since the API of passing Zz was never fully agreed upon.

Do we want to go with -v /foo/bar:rwZ or -v /foo/bar:rw:Z

crosbymichael commented:
I think `-v /foo/bar:rwZ` if it's a non breaking change

cpuguy83 commented:
Yes, I like the rwZ syntax.
I think it would be best to keep the parsing of this in docker/daemon.go and add a new field to the container struct with the label type (private/shared) for each volume and that can be applied directly when configuring the mounts being sent to libcontainer (in `setupMounts()`).

rhatdan commented:
Well if that is something @crosbymichael wants I can prepare a patch for it.

SvenDowideit commented:
I just built this PR's code and put it on an RHEL7.1 instance I'm using to play with and learn SElinux, and it has some 'interesting' failures:

```
[root@RHEL71 ~]# docker run --rm -it  -v /var/run/docker.sock:/var/run/docker.sock ubuntu
root@1f2b76091aa0:/# ping docker.com
ping: unknown host docker.com
root@1f2b76091aa0:/# more /etc/resolv.conf
/etc/resolv.conf: Permission denied
root@1f2b76091aa0:/# exit
[root@RHEL71 ~]#
[root@RHEL71 ~]# docker run --rm -it  -v /var/run/docker.sock:/var/run/docker.sock:wrZ --net=host  ubuntu
Timestamp: 2015-03-06 14:03:28.352902326 +1100 AEDT
Code: System error

Message: invalid argument

Frames:
---
0: setupRootfs
Package: github.com/docker/libcontainer
File: rootfs_linux.go@51
---
1: Init
Package: github.com/docker/libcontainer.(*linuxStandardInit)
File: standard_init_linux.go@51
---
2: StartInitialization
Package: github.com/docker/libcontainer.(*LinuxFactory)
File: factory_linux.go@203
---
3: initializer
Package: github.com/docker/docker/daemon/execdriver/native
File: init.go@35
---
4: Init
Package: github.com/docker/docker/pkg/reexec
File: reexec.go@26
---
5: main
Package: main
File: docker.go@28
---
6: main
Package: runtime
File: proc.go@63
---
7: goexit
Package: runtime
File: asm_amd64.s@2232
FATA[0000] Error response from daemon: Cannot start container a9ca7df4c2cbf3d87d10ecddd63df66c20f2109f4eee2590b532e7476160135f: [8] System error: invalid argument
[root@RHEL71 ~]#
```

vbatts commented:
That looks like the libcontainer mqueue failure.
On Mar 12, 2015 9:06 PM, "Sven Dowideit" <notifications@github.com> wrote:

> I just built this PR's code and put it on an RHEL7.1 instance I'm using to
> play with and learn SElinux, and it has some 'interesting' failures:
>
> [root@RHEL71 ~]# docker run --rm -it  -v /var/run/docker.sock:/var/run/docker.sock ubuntu
> root@1f2b76091aa0:/# ping docker.com
> ping: unknown host docker.com
> root@1f2b76091aa0:/# more /etc/resolv.conf
> /etc/resolv.conf: Permission denied
> root@1f2b76091aa0:/# exit
> [root@RHEL71 ~]#
> [root@RHEL71 ~]# docker run --rm -it  -v /var/run/docker.sock:/var/run/docker.sock:wrZ --net=host  ubuntu
> Timestamp: 2015-03-06 14:03:28.352902326 +1100 AEDT
> Code: System error
>
> Message: invalid argument
>
> Frames:
> ---
> 0: setupRootfs
> Package: github.com/docker/libcontainer
> File: rootfs_linux.go@51
> ---
> 1: Init
> Package: github.com/docker/libcontainer.(*linuxStandardInit)
> File: standard_init_linux.go@51
> ---
> 2: StartInitialization
> Package: github.com/docker/libcontainer.(*LinuxFactory)
> File: factory_linux.go@203
> ---
> 3: initializer
> Package: github.com/docker/docker/daemon/execdriver/native
> File: init.go@35
> ---
> 4: Init
> Package: github.com/docker/docker/pkg/reexec
> File: reexec.go@26
> ---
> 5: main
> Package: main
> File: docker.go@28
> ---
> 6: main
> Package: runtime
> File: proc.go@63
> ---
> 7: goexit
> Package: runtime
> File: asm_amd64.s@2232
> FATA[0000] Error response from daemon: Cannot start container a9ca7df4c2cbf3d87d10ecddd63df66c20f2109f4eee2590b532e7476160135f: [8] System error: invalid argument
> [root@RHEL71 ~]#
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/5910#issuecomment-78727689>.
>

SvenDowideit commented:
mmm, I'm going to run this test again - I'm suddenly wondering if the above was with the master.dockerproject.org binary, or the one i built from the PR.

SvenDowideit commented:
ah, post reboot, I get 

```
[root@RHEL71 ~]# docker version
Client version: 1.5.0-dev
Client API version: 1.18
Go version (client): go1.4.2
Git commit (client): 681115a
OS/Arch (client): linux/amd64
Server version: 1.5.0-dev
Server API version: 1.18
Go version (server): go1.4.2
Git commit (server): 681115a
OS/Arch (server): linux/amd64
[root@RHEL71 ~]# docker run --rm -it hello-world
Timestamp: 2015-03-06 14:31:13.930951683 +1100 AEDT
Code: System error

Message: invalid argument

Frames:
---
0: setupRootfs
Package: github.com/docker/libcontainer
File: rootfs_linux.go@51
---
1: Init
Package: github.com/docker/libcontainer.(*linuxStandardInit)
File: standard_init_linux.go@51
---
2: StartInitialization
Package: github.com/docker/libcontainer.(*LinuxFactory)
File: factory_linux.go@203
---
3: initializer
Package: github.com/docker/docker/daemon/execdriver/native
File: init.go@35
---
4: Init
Package: github.com/docker/docker/pkg/reexec
File: reexec.go@26
---
5: main
Package: main
File: docker.go@28
---
6: main
Package: runtime
File: proc.go@63
---
7: goexit
Package: runtime
File: asm_amd64.s@2232
FATA[0000] Error response from daemon: Cannot start container 739f7ca9a30bea07bb90d4d07962850b45cafcabd8659b17bb7c3214a7746f28: [8] System error: invalid argument
[root@RHEL71 ~]#
```

so at least its mildly consistent.

rhatdan commented:
@SvenDowideit THat is not the way it works.  If you want to try this out, create content in /var/lib/test

And then use

docker run -ti --rm -v /var/lib/test:/var/lib/test:Z fedora ls -dZ /var/lib/test
docker run -ti --rm -v /var/lib/test:/var/lib/test:z fedora ls -dZ /var/lib/test

You should see different labels.  Running with the /run/docker.sock inside the container will never work with SELinux in a non privileged container.

ibuildthecloud commented:
@crosbymichael @rhatdan I actually vote for `-v /foo/bar:rw:Z`.  The reason is that I'd like to add shared and slave mount support to Docker.  The syntax `-v /foo/bar:rw:Z:shared` looks better than `-v /foo/bar:rwZshared`.  The option without the `:` only works if we have a short form of all options.  I guess we should invent s and S for shared/slave, but I'd rather not.

cpuguy83 commented:
Hmm, every new ":" makes the UI (and the parsing) that much worse.

rhatdan commented:
I think we had the ":" debate and it lost.  I believe Single Letter Options are best.  I would prefer to just use  r,w,z,Z and ignore o.  
rw -> w for backwards compatibility
ro -> r for backwards compatibility

Eventually I predict we will have u and maybe g for changing the ownership to default UID and GID.

rhatdan commented:
I have no idea why this is failing on the Windows build.  But I don't think my patch is involved.

vbatts commented:
@rhatdan i've just issued a rebuild for windows. Hopefully that's all it needs.

tiborvass commented:
@rhatdan something's wrong with the commits, they are merge commits. Would you be able to fix it?

rhatdan commented:
@tiborvass Fixed

ibuildthecloud commented:
Last comment and then I'll shut up.  From a coding perspective "rwZ" or "rw:Z" makes almost no difference.  You can easily parse both and if your super lazy just strip ":" from "rw:Z" to become "rwZ".

What it does impact is the user experience.  So ro, rw are super obvious to users.  z and Z are bit more obscure but if you need to set them they you probably already understand z and Z.  But for other mount options like shared and slave (which I want to add) picking sS or sh/sl is pretty obscure.  It would be nice if I had a syntax that easily allowed longer form options.

rhatdan commented:
What does shared/slave mean?

jfrazelle commented:
I think the design is good now and implementation is ok so LGTM

ping @crosbymichael

cpuguy83 commented:
So currently `volumes-from` will set a previously shared mount to private, and take over any private mounts so the original container wouldn't be able to access the files... and this gets reset any time either the producer or the consumer is created.

```bash
~: docker run -v /test:/test:z /test2:/test2:Z -d --name test busybox top
~: ls -lhZ /

# <results filtered>
drwxr-xr-x. root root system_u:object_r:svirt_sandbox_file_t:s0 test
drwxr-xr-x. root root system_u:object_r:svirt_sandbox_file_t:s0:c576,c842 test2

~: docker run -d --volumes-from test busybox
~: ls -lhZ /

# <results filtered>
drwxr-xr-x. root root system_u:object_r:svirt_sandbox_file_t:s0:c73,c466 test
drwxr-xr-x. root root system_u:object_r:svirt_sandbox_file_t:s0:c73,c466 test2

~: docker restart test
~: ls -lhZ

# <results filtered>
drwxr-xr-x. root root system_u:object_r:svirt_sandbox_file_t:s0 test
drwxr-xr-x. root root system_u:object_r:svirt_sandbox_file_t:s0:c576,c842 test2
```

So basically, volumes-from pretty much breaks this completely.
I would say we should not set relabling when using volumes-from, this way the labels stay as is.

Can we set a default value when none is provided? Right now if neither z or Z are provided, the label is `unconfined_u:object_r:default_t:s0`... as I'm sure is the case in released versions of docker and generally horrible behavior since users have to manually relabel their volumes.
I know normal volumes (non binds) get `object_r:svirt_sandbox_file_t:s0`

rhatdan commented:
Yes volumes-from should not be changing the label, or should change it to a shared label, since potentially two or more containers are sharing the content.

default_t is the default type of directories created in /, and should not be used by any confined domain.

I don't want to change any labels unless an admin tells docker to change it.  Creating a docker container volumes being the exception, since docker already owns the content in /var/lib/docker/vfs.

rhatdan commented:
Fixed the --volumes-from option to not relabel.

cpuguy83 commented:
@rhatdan Just tested again and still seeing the same thing with `volumes-from`

rhatdan commented:
@cpuguy83 Seems to work for me.  Did you recreate /test and /test2 and check the labels before doing the volumesfrom?

cpuguy83 commented:
Yes, this was on a brand-new host, and I don't see in the code where it would be skipping on volumes-from.

rhatdan commented:
@cpuguy83 Are you happy with the current patch?  It should not effect the volumes-from either.

LK4D4 commented:
@crosbymichael Would you mind to review?
@rhatdan Rebase is needed.

tianon commented:
On the `:something` UI debate, why not match the already-familiar UI of `mount -o`? (comma-separated)

`-v /a:/b:rw,uid=1000,gid=1000,shared` is a lot easier to parse (visually and programmatically) than colons, and gives flexibility for arguments and things that don't compress to a single character well.

LK4D4 commented:
@tianon Hell, returning UX back :)

rhatdan commented:
I am fine with comma separated, just want someone to pick something.  As this patch set approaches one year out.

tianon commented:
@lk4d4 ALWAYS

rhatdan committed:
Modify volume mounts SELinux labels on the fly based on :Z or :z

This patch is extending the qualifiers on the -v command to allow
an admin to tell the system to relabel, content. There might be a
need for something similar for changing the DAC Permissions.

Docker-DCO-1.1-Signed-off-by: Dan Walsh <dwalsh@redhat.com> (github: rhatdan)

tiborvass commented:
I'm +1 on @tianon's comma separated options.

LK4D4 commented:
Let's do comma-separated and merge it.

LK4D4 commented:
ping @rhatdan

rhatdan commented:
@LK4D4 If I switch to "," separated list and someone else changes it, my head will explode.

rhatdan commented:
Now switched to "," separated list.  Might need to bring this out in the docs a little better.  I did add a NOte in man page.

crosbymichael commented:
@rhatdan this one is good to go now right?  No more changes in libcontainer or this Pr?

thaJeztah commented:
Looks like there's a merge commit in the PR

rhatdan commented:
@crosbymichael There is a libcontainer patch which would allow us to remove the label() call from this patch set.

https://github.com/docker/libcontainer/pull/509

When this is merged into docker we can simplify this patch a little.  Remove the relabel() function.

calavera commented on daemon/volumes.go:
```diff
@@ -163,11 +165,29 @@ func (container *Container) parseVolumeMountConfig() (map[string]*Mount, error)
 			return nil, fmt.Errorf("Duplicate volume %q: %q already in use, mounted from %q", path, mountToPath, m.volume.Path)
 		}
+
+		runtime.LockOSThread()
```
Is there any reason to lock the thread here?
It also needs a `defer runtime.UnlockOSThread()`.

rhatdan replied:
Yes SELinux labeling sets up a per process/thread label,  This is basically saying if this thread creates any content on disk label it with mountlabel.

I did not know UnlockOSThread needed to be called, but no problem adding it.

LK4D4 replied:
Also, note for @jfrazelle . You should change label back in this thread.

jfrazelle commented:
hey @rhatdan I'll carry this so we can get it in for 1.7

jfrazelle commented:
carried in #12572
